File: 12.Search_In_Rotated_Sorted_Array/sol.py
import os
import sys

# ======================================================================
# 🧠 CHALLENGE: Search in Rotated Sorted Array (Python Version)
# ======================================================================
# Description:
# There is an integer array `nums` sorted in ascending order (with distinct values).
# Prior to being passed to your function, `nums` is possibly rotated at an
# unknown index k.
#
# Given the array `nums` after the possible rotation and an integer `target`,
# return the index of `target` if it is in `nums`, or -1 if it is not.
#
# 📋 Rules:
# 1. You must write an algorithm with O(log n) runtime complexity.
# 2. Return the index of the target, or -1.
#
# 💡 Examples:
# - search([4,5,6,7,0,1,2], 0) => 4
# - search([4,5,6,7,0,1,2], 3) => -1
# - search([1], 0)             => -1
# ======================================================================

# region [📚 Reference Solutions] (Solutions hidden as requested)
# (Focus on implementing your own logic in the Practice Area below!)
# endregion


# ======================================================================
#  region [✍️ Practice Area]
#  Please write your solution between the markers below.
# ======================================================================
# <PRACTICE_START>
def search(nums, target):
    # TODO: Implement your solution here.
    return -1


# <PRACTICE_END>
# endregion

# ======================================================================
#  region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
# ======================================================================


def reset_practice_area():
    print("\n🔄 Resetting Practice Area to default state...")
    MARKER_START = "# <PRACTICE_" + "START>"
    MARKER_END = "# <PRACTICE_" + "END>"

    default_code_lines = [
        "def search(nums, target):\n",
        "    # TODO: Implement your solution here.\n",
        "    return -1\n",
    ]

    try:
        current_file = __file__
        with open(current_file, "r", encoding="utf-8") as f:
            lines = f.readlines()

        start_idx = -1
        end_idx = -1

        for i, line in enumerate(lines):
            if MARKER_START in line:
                start_idx = i
            if MARKER_END in line:
                end_idx = i

        if start_idx == -1 or end_idx == -1 or start_idx >= end_idx:
            print("⚠️ Error: Markers not found. Reset cancelled.")
            return

        new_content = lines[: start_idx + 1] + default_code_lines + lines[end_idx:]

        with open(current_file, "w", encoding="utf-8") as f:
            f.writelines(new_content)
        print("✨ Reset complete! The file is ready for a fresh start.")
    except Exception as e:
        print(f"⚠️ Reset failed: {e}")


if __name__ == "__main__":
    test_cases = [
        ([4, 5, 6, 7, 0, 1, 2], 0, 4),
        ([4, 5, 6, 7, 0, 1, 2], 3, -1),
        ([1], 0, -1),
        ([1], 1, 0),
        ([3, 1], 1, 1),
        ([5, 1, 3], 5, 0),
        ([4, 5, 6, 7, 8, 1, 2, 3], 8, 4),
    ]

    print(f"\n🧪 Testing your [search] function...\n")

    header = f"{'Input nums':<25} | {'Target':<6} | {'Expected':<10} | {'Actual':<10} | Status"
    print(header)
    print("-" * len(header))

    all_pass = True
    for nums, target, expected in test_cases:
        try:
            result = search(nums, target)
        except Exception as e:
            result = "Error"

        is_match = result == expected
        status_icon = "✅ PASS" if is_match else "❌ FAIL"
        if not is_match:
            all_pass = False

        # Format output
        nums_str = str(nums)
        if len(nums_str) > 23:
            nums_str = nums_str[:20] + "..."

        print(
            f"{nums_str:<25} | {str(target):<6} | {str(expected):<10} | {str(result):<10} | {status_icon}"
        )

    print("-" * len(header))
    if all_pass:
        print("\n🎉 Fantastic! All test cases passed.")
        reset_practice_area()
    else:
        print("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")

# endregion

File: 12.Search_In_Rotated_Sorted_Array/sol.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"strings"
)

// ======================================================================
// 🧠 CHALLENGE: Search in Rotated Sorted Array (Go Version)
// ======================================================================
// Description:
// There is an integer array `nums` sorted in ascending order (with distinct values).
// Prior to being passed to your function, `nums` is possibly rotated at an
// unknown index k.
//
// Given the array `nums` after the possible rotation and an integer `target`,
// return the index of `target` if it is in `nums`, or -1 if it is not.
//
// 📋 Rules:
// 1. You must write an algorithm with O(log n) runtime complexity.
// 2. Return the index of the target, or -1.
//
// 💡 Examples:
// - Search([4,5,6,7,0,1,2], 0) => 4
// - Search([4,5,6,7,0,1,2], 3) => -1
// - Search([1], 0)             => -1
// ======================================================================

// #region [📚 Reference Solutions] (Solutions hidden as requested)
// (Focus on implementing your own logic in the Practice Area below!)
// #endregion

// ======================================================================
//
//	#region [✍️ Practice Area]
//	Please write your solution between the markers below.
//
// ======================================================================
// <PRACTICE_START>
func Search(nums []int, target int) int {
	// TODO: Implement your solution here.
	return -1
}

// <PRACTICE_END>
// #endregion

// ======================================================================
//  #region [🚀 Test Runner & Auto-Reset] (Do not modify below this line)
// ======================================================================

func main() {
	runTests()
}

type TestCase struct {
	nums     []int
	target   int
	expected int
}

func runTests() {
	testCases := []TestCase{
		{[]int{4, 5, 6, 7, 0, 1, 2}, 0, 4},
		{[]int{4, 5, 6, 7, 0, 1, 2}, 3, -1},
		{[]int{1}, 0, -1},
		{[]int{1}, 1, 0},
		{[]int{3, 1}, 1, 1},
		{[]int{5, 1, 3}, 5, 0},
		{[]int{4, 5, 6, 7, 8, 1, 2, 3}, 8, 4},
	}

	fmt.Printf("\n🧪 Testing your [Search] function...\n\n")

	header := fmt.Sprintf("%-25s | %-6s | %-10s | %-10s | Status", "Input nums", "Target", "Expected", "Actual")
	fmt.Println(header)
	fmt.Println(strings.Repeat("-", len(header)))

	allPass := true

	for _, tc := range testCases {
		result := Search(tc.nums, tc.target)

		isMatch := result == tc.expected
		statusIcon := "✅ PASS"
		if !isMatch {
			statusIcon = "❌ FAIL"
			allPass = false
		}

		// Helper to format slices
		fmtSlice := func(s []int) string {
			return fmt.Sprintf("%v", s)
		}

		numsStr := fmtSlice(tc.nums)
		if len(numsStr) > 23 {
			numsStr = numsStr[:20] + "..."
		}

		fmt.Printf("%-25s | %-6d | %-10v | %-10v | %s\n",
			numsStr, tc.target, tc.expected, result, statusIcon)
	}

	fmt.Println(strings.Repeat("-", len(header)))

	if allPass {
		fmt.Println("\n🎉 Fantastic! All test cases passed.")
		resetPracticeArea()
	} else {
		fmt.Println("\n⚠️  Some tests failed. Keep trying! (The file will not reset yet)")
	}
}

func resetPracticeArea() {
	fmt.Println("\n🔄 Resetting Practice Area to default state...")

	markerStart := "// <PRACTICE_" + "START>"
	markerEnd := "// <PRACTICE_" + "END>"

	defaultCode := []string{
		"func Search(nums []int, target int) int {",
		"\t// TODO: Implement your solution here.",
		"\treturn -1",
		"}",
	}

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		fmt.Println("⚠️ Error: Could not determine file path. Reset cancelled.")
		return
	}

	content, err := os.ReadFile(currentFile)
	if err != nil {
		fmt.Printf("⚠️ Error reading file: %v\n", err)
		return
	}

	lines := strings.Split(string(content), "\n")
	startIdx := -1
	endIdx := -1

	for i, line := range lines {
		if strings.Contains(line, markerStart) {
			startIdx = i
		} else if strings.Contains(line, markerEnd) {
			endIdx = i
		}
	}

	if startIdx == -1 || endIdx == -1 || startIdx >= endIdx {
		fmt.Println("⚠️ Error: Markers not found or invalid. Reset cancelled.")
		return
	}

	newLines := make([]string, 0)
	newLines = append(newLines, lines[:startIdx+1]...)
	newLines = append(newLines, defaultCode...)
	newLines = append(newLines, lines[endIdx:]...)

	output := strings.Join(newLines, "\n")
	err = os.WriteFile(currentFile, []byte(output), 0644)
	if err != nil {
		fmt.Printf("⚠️ Error writing file: %v\n", err)
		return
	}

	fmt.Println("✨ Reset complete! The file is ready for a fresh start.")
}

// #endregion
